### DISCUSSION TITLE
Error on iOS but not web: "Invariant Violation: `new NativeEventEmitter()` requires a non-null argument., js engine: hermes"

### DISCUSSION THREAD

**Discussion by voxoid0** [invertase/react-native-firebase#8102] [Q&A] [answered]
Hi, I have an expo app in Bare workflow, and react-native-firebase (app, auth, db) works fine on web, but on iOS I get a runtime error that prevents the app from running:

`Invariant Violation: `new NativeEventEmitter()` requires a non-null argument., js engine: hermes`

I've tried the various solutions across the web for that error, with no luck, and also followed all the instructions for iOS at https://rnfirebase.io/ , tried lots of cleaning and rebuilding (including `pod deintegrate`, `npx pod-install`, `npm cache clean --force`, clean and rebuild in XCode. I also checked for version compatibility between iOS target, react-native, and react-native-firebase, and I *think* it looks OK...plus I think peer dependencies should typically handle that. Below is my package.json and Podfile; I'm using XCode 15.4, iOS target 13.4, react-native-firebase 21.2.0, react-native 0.74.5.

```
{
  "name": "redacted",
  "version": "1.0.0",
  "type": "module",
  "main": "expo-router/entry",
  "scripts": {
    "start": "expo start",
    "reset-project": "node ./scripts/reset-project.js",
    "android": "expo run:android",
    "ios": "expo run:ios",
    "web": "expo start --web",
    "build": "tsc",
    "test": "npm run build && jest --watchAll",
    "lint": "expo lint"
  },
  "dependencies": {
    "@expo/metro-runtime": "~3.2.3",
    "@expo/vector-icons": "^14.0.2",
    "@mediapipe/tasks-vision": "^0.10.17",
    "@motiz88/react-native-midi": "^0.0.6",
    "@react-native-async-storage/async-storage": "1.23.1",
    "@react-native-firebase/app": "^21.2.0",
    "@react-native-firebase/auth": "^21.2.0",
    "@react-native-firebase/crashlytics": "^21.2.0",
    "@react-native-firebase/firestore": "^21.2.0",
    "@react-navigation/native": "^6.0.2",
    "@react-navigation/stack": "^6.4.1",
    "@react-three/drei": "^9.114.4",
    "@react-three/fiber": "^8.17.10",
    "@techstark/opencv-js": "^4.10.0-release.1",
    "@tonaljs/tonal": "^4.10.0",
    "@tweenjs/tween.js": "^25.0.0",
    "@types/graphlib": "^2.1.12",
    "@types/lodash": "^4.17.10",
    "@types/luxon": "^3.4.2",
    "@types/three": "^0.169.0",
    "@types/yaml": "^1.9.7",
    "cross-fetch": "^4.0.0",
    "expo": "~51.0.37",
    "expo-av": "^14.0.7",
    "expo-constants": "~16.0.2",
    "expo-dev-client": "~4.0.28",
    "expo-firebase-analytics": "^8.0.0",
    "expo-font": "~12.0.9",
    "expo-gl": "~14.0.2",
    "expo-linking": "~6.3.1",
    "expo-router": "~3.5.23",
    "expo-speech-recognition": "^0.2.22",
    "expo-splash-screen": "~0.27.5",
    "expo-status-bar": "~1.12.1",
    "expo-system-ui": "~3.0.7",
    "expo-web-browser": "~13.0.3",
    "firebase": "^11.0.1",
    "formik": "^2.4.6",
    "graphlib": "^2.1.8",
    "lodash": "^4.17.21",
    "luxon": "^3.5.0",
    "metro": "^0.81.0",
    "midifile-ts": "^1.5.1",
    "mirada": "^0.0.15",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "react-native": "0.74.5",
    "react-native-async-storage": "^0.0.1",
    "react-native-gesture-handler": "~2.16.1",
    "react-native-keyboard-aware-scroll-view": "^0.9.5",
    "react-native-paper": "^5.12.5",
    "react-native-quick-base64": "^2.1.2",
    "react-native-reanimated": "~3.10.1",
    "react-native-safe-area-context": "4.10.5",
    "react-native-screens": "3.31.1",
    "react-native-vision-camera": "^4.5.3",
    "react-native-web": "~0.19.10",
    "react-native-worklets-core": "^1.3.3",
    "react-navigation": "^5.0.0",
    "socket.io-client": "^4.8.0",
    "three": "^0.166.1",
    "undefined": "react-native-async-storage/async-storage",
    "yaml": "^2.5.1",
    "yup": "^1.4.0"
  },
  "devDependencies": {
    "@babel/core": "^7.20.0",
    "@babel/preset-env": "^7.25.7",
    "@babel/preset-typescript": "^7.25.7",
    "@types/jest": "^29.5.13",
    "@types/react": "~18.2.45",
    "@types/react-test-renderer": "^18.0.7",
    "babel-jest": "^29.7.0",
    "babel-plugin-module-resolver": "^5.0.2",
    "jest": "^29.2.1",
    "jest-expo": "~51.0.3",
    "react-test-renderer": "18.2.0",
    "ts-jest": "^29.2.5",
    "ts-mockito": "^2.6.1",
    "typescript": "~5.3.3"
  },
  "private": true
}
```

```
# Override Firebase SDK Version (From instructions at https://rnfirebase.io/)
$FirebaseSDKVersion = '11.4.0'

require File.join(File.dirname(`node --print "require.resolve('expo/package.json')"`), "scripts/autolinking")
require File.join(File.dirname(`node --print "require.resolve('react-native/package.json')"`), "scripts/react_native_pods")

require 'json'
podfile_properties = JSON.parse(File.read(File.join(__dir__, 'Podfile.properties.json'))) rescue {}

ENV['RCT_NEW_ARCH_ENABLED'] = podfile_properties['newArchEnabled'] == 'true' ? '1' : '0'
ENV['EX_DEV_CLIENT_NETWORK_INSPECTOR'] = podfile_properties['EX_DEV_CLIENT_NETWORK_INSPECTOR']

use_autolinking_method_symbol = ('use' + '_native' + '_modules!').to_sym
origin_autolinking_method = self.method(use_autolinking_method_symbol)
self.define_singleton_method(use_autolinking_method_symbol) do |*args|
  if ENV['EXPO_UNSTABLE_CORE_AUTOLINKING'] == '1'
    Pod::UI.puts('Using expo-modules-autolinking as core autolinking source'.green)
    config_command = [
      'node',
      '--no-warnings',
      '--eval',
      'require(require.resolve(\'expo-modules-autolinking\', { paths: [require.resolve(\'expo/package.json\')] }))(process.argv.slice(1))',
      'react-native-config',
      '--json',
      '--platform',
      'ios'
    ]
    origin_autolinking_method.call(config_command)
  else
    origin_autolinking_method.call()
  end
end

platform :ios, podfile_properties['ios.deploymentTarget'] || '13.4'
install! 'cocoapods',
  :deterministic_uuids => false

prepare_react_native_project!

target 'encordiapiano' do
  use_expo_modules!
  config = use_native_modules!

  use_frameworks! :linkage => podfile_properties['ios.useFrameworks'].to_sym if podfile_properties['ios.useFrameworks']
  use_frameworks! :linkage => ENV['USE_FRAMEWORKS'].to_sym if ENV['USE_FRAMEWORKS']

  use_frameworks! :linkage => :static
  $RNFirebaseAsStaticFramework = true

  use_react_native!(
    :path => config[:reactNativePath],
    :hermes_enabled => podfile_properties['expo.jsEngine'] == nil || podfile_properties['expo.jsEngine'] == 'hermes',
    # An absolute path to your application root.
    :app_path => "#{Pod::Config.instance.installation_root}/..",
    :privacy_file_aggregation_enabled => podfile_properties['apple.privacyManifestAggregationEnabled'] != 'false',
  )

  pod 'FirebaseCore', :modular_headers => true
  # Add the FirebaseCoreInternal pod with modular headers
#   pod 'GoogleUtilities', :modular_headers => true

  post_install do |installer|
    react_native_post_install(
      installer,
      config[:reactNativePath],
      :mac_catalyst_enabled => false,
      :ccache_enabled => podfile_properties['apple.ccacheEnabled'] == 'true',
    )

    # This is necessary for Xcode 14, because it signs resource bundles by default
    # when building for devices.
    installer.target_installation_results.pod_target_installation_results
      .each do |pod_name, target_installation_result|
      target_installation_result.resource_bundle_targets.each do |resource_bundle_target|
        resource_bundle_target.build_configurations.each do |config|
          config.build_settings['CODE_SIGNING_ALLOWED'] = 'NO'
        end
      end
    end
  end

  post_integrate do |installer|
    begin
      expo_patch_react_imports!(installer)
    rescue => e
      Pod::UI.warn e
    end
  end
end
```

Please help? 🥺🐶🙏

**Comment by mikehardy** [comment]
Hi there!

> I've tried the various solutions across the web for that error, with no luck

That's vague, I'm not sure what you tried or did not try and what the results of each experiment were, whether you reverted after the experiment etc. Science (and troubleshooting is a science...) needs to be methodical, share methodology and results etc.

In the past I have never had a single person provide me a reproduction of this, as you've likely read if you read through the related issues: https://github.com/invertase/react-native-firebase/issues?q=NativeEventEmitter+requires+a+non-null+argument

Lots of "me too", no "here's a reproduction", so it is hard to help.

For the absolute avoidance of doubt, a correctly integrated react-native-firebase installation does not have this problem, and I provide an automated reproduction that proves it: https://github.com/mikehardy/rnfbdemo/blob/main/make-demo.sh - so there is a high bar here to prove a claim that "react-native-firebase does not load at all", as I show it loading.

What's the difference between that build proof and your project? Expo. I don't use Expo so I can't help much there but the common thread on those other issues is "I'm using Expo managed..." and that perhaps a prebuild will fix it.

All I can say with certainty is: this repository does not currently have any reproduction that shows this problem, it only has reproductions that show it does not have it.

**Comment by voxoid0** [comment]
Hi Mike, thanks for the quick reply, yeah I assumed it's likely not a bug with any of the libs or SDKs so I'm not surprised that a bare minimum react-native-firebase app works, just a setup/configuration issue that I was hoping someone in the community may have deep exerience enough to know how to track it down or what else to check. Yeah `prebuild` is another thing I tried, forgot to list that one; I also checked library code for usage of the NativeEventEmitter() constructor that doesn't pass a parameter. I was hoping I wouldn't have to go through a long process like starting with the working minimum and slowly add parts of my app. I'll keep poking around and see what differences I can figure out.

**Comment by voxoid0** [comment]
@mikehardy do you have any helpful insight as to what tends to cause this in general and how to troubleshoot it? E.g. my best guess as a somewhat-beginner on the platform is library incompatibility, since it sounds like there is code calling `new NativeEventEmitter()` without an argument, while a different version of react-native expects an argument. OR probably more likely is that the `nativeModule` argument is being passed null which, according to the attached stack trace, is gotten by `getReactNativeModule('RNFBAppModule')`, so perhaps the key is why that would be returning null?:

```
class RNFBNativeEventEmitter extends NativeEventEmitter {
  constructor() {
    super(getReactNativeModule('RNFBAppModule'));
    this.ready = false;
  }
```

![PXL_20241031_121023514](https://github.com/user-attachments/assets/a0e54d84-c20d-4c38-a57d-bdf7b0317c66)

**Reply by mikehardy** [reply]
I got this once yesterday when I tried to load the rnfbdemo app javascript bundle into an native example app built for a different repo, there was no app module initialized in the native example app since it wasn't a react-native-firebase app, so the javascript code couldn't find the default firebase app and I suppose you are right - in that case it's null

So my guess is that people stumbling in to this problem have not set up their applications correctly at the native level to do the initialization of firebase so there is no native app

You might carefully check that these steps were followed https://rnfirebase.io/#configure-firebase-with-ios-credentials

**Reply by mikehardy** [reply]
I proposed #8106 to improve the developer experience here

**Comment by voxoid0** [comment]
I rechecked everything carefully against the iOS instructions at rnfirebase.io, and also checked the XCode project against Firebase's own iOS instructions. Trying to figure out *why* there is no RNFBAppModule; is there a way to turn on debug level logging, like maybe at the Objective-C level in AppDelegate.mm before [FIRApp configure] is called?

Chat-GPT 4o suggested `[[NSUserDefaults standardUserDefaults] setBool:YES forKey:@"RNFBDebug"];`, but it's probably hallucinating.

```
- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  [[NSUserDefaults standardUserDefaults] setBool:YES forKey:@"RNFBDebug"];
  [FIRApp configure];
```

**Reply by voxoid0** [reply]
Oh I think I found log level: `[FIRConfiguration sharedInstance].loggerLevel = FIRLoggerLevelMax;`

**Reply by mikehardy** [reply]
https://firebase.google.com/docs/reference/ios/firebasecore/api/reference/Classes/FIRApp

> Enable debug mode by passing the -FIRDebugEnabled argument to the application. You can add this argument in the application’s Xcode scheme. When debug mode is enabled via -FIRDebugEnabled, further executions of the application will also be in debug mode. In order to return to default mode, you must explicitly disable the debug mode with the application argument -FIRDebugDisabled.
> 
> It is also possible to change the default logging level in code by calling FirebaseConfiguration.shared.setLoggerLevel(_:) with the desired level.

**Reply by mikehardy** [reply]
it may be that the GoogleServices-Info.plist is not included correctly in the build ?

**Comment by voxoid0** [comment, marked as answer]
In my case, the only blunder was that I still had some packages installed from when had started out using the JS package instead of @react-native-firebase packages. My original packages.json above has packages `firebase` and `expo-firebase-analytics`, and after removing them, then deintegrating and installing the pods, cleaning and building the app, my bare expo app no longer gets the NativeEventEmitter() error. I don't know exactly why that causes the RNFBApp module to be unavailable, but apparently there is interference between those packages.

Thanks for the support, @mikehardy.

**Reply by mikehardy** [reply]
Happy to help - sounds like that was a tough one, I'll add "conflicting packages" to the mental list of things that can go wrong! Good luck with your project

**Comment by brunogdsdev** [comment]
To me one lib was conflicting and I had to increase the value of deployment and run 'pod Install'. This worked for me

1-`platform :ios, podfile_properties['ios.deploymentTarget'] || '18.4'`
2-remove ios/Pod and Podfile.lock
3-`pod Install`
